drivers: serial: nrf: fix compilation when system clock is disabled

mspeder opened the pull request:
When SYS_CLOCK_EXISTS is disabled, k_timers functions are not available. This fix simply removes all those calls mainly by using a bunch of #ifdef. In that case, obviously, timeout values are simply ignored, but this is the expected behavior anyway, as mentioned in the SYS_CLOCK_EXISTS documentation. If SYS_CLOCK_EXISTS is enabled (which is generally the case), this fix does not change previous behavior.

Signed-off-by: Matthieu Speder <mspeder@users.sourceforge.net>

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -73,8 +73,10 @@ static struct {
 	size_t rx_secondary_buffer_length;
 	volatile size_t rx_counter;
 	volatile size_t rx_offset;
+#ifdef CONFIG_SYS_CLOCK_EXISTS
 	int32_t rx_timeout;
 	struct k_timer rx_timeout_timer;
+#endif
```
This is not a proper way of solving the compilation issue. Without timers, the asynchronous API will not be handled correctly in the nRF UART drivers, so users should not be able to enable this API in such case.
Instead of adding all these `#ifdef`s, you should rather change: https://github.com/zephyrproject-rtos/zephyr/blob/75a2ef8fc02a417f0d26307105f80b136a666dc8/drivers/serial/Kconfig.nrfx#L11 to:
```
	select SERIAL_SUPPORT_ASYNC if SYS_CLOCK_EXISTS
```

nordic-krch replied:
@anangl yes, api will have limited use, on tx side there will be no timeout when hwfc is enabled. On async there will be no events on partial data. Only when full buffer is received. Why not use interrupt driven api then if you have no sys clock? What would be the benefits of async api?

mspeder replied:
My use case is the following one : uart is receiving packets of data and so I perfectly know the expected size I will receive. So in that case there is no need to have partial data. I only need to be notified when the full message is received. Using interrupt api will simply consume cpu power with dozens of intermediate interrupts (total size/hardware fifo size), instead of relying on hardware doing the job for the full packet (which is the whole point of easy dma I think).

So I think there is a usage for DMA-driven async uarte without the need of any timeout.
The behavior is the same when you specify SYS_FOREVER_MS as timeout value, so this looked an anticipated behavior anyway.

mspeder replied:
@nordic-krch I agree that the commit looks a bit scary like this (it is better when you look at the complete file though). Indeed I tried my best to keep the existing style. Here is sample extract from existing code :

```
/* Device data structure */
struct uarte_nrfx_data {
	const struct device *dev;
	struct uart_config uart_config;
#ifdef UARTE_INTERRUPT_DRIVEN
	struct uarte_nrfx_int_driven *int_driven;
#endif
#ifdef CONFIG_UART_ASYNC_API
	struct uarte_async_cb *async;
#endif
	atomic_val_t poll_out_lock;
#ifdef CONFIG_PM_DEVICE
	uint32_t pm_state;
#endif
	uint8_t char_out;
	uint8_t rx_data;
	gppi_channel_t ppi_ch_endtx;
};
```

Pattern seemed very similar. And since we are targeting embedded device with limited RAM I guess it still makes sense though ?

Feedback about nrfx_glue is a very good point, as this specific part of code has a broader scope than just uart.
I will split it as soon as there is a consensus on the global usefulness of async without timeouts.

As this is my first contribution to zephyr I am not sure how to properly create a test case. I will try to figure out.

anangl replied:
@mspeder I see that I misunderstood the intention of this PR in the first place. Yes, the use case you provided is reasonable (perhaps it would be good to mention it in the commit message). I agree with @nordic-krch that that many `#ifdef`s is not nice looking at all, but I guess we'll need to live with it. I'd only suggest to use a definition like:
```
#define TIMEOUTS_AVAILABLE  IS_ENABLED(CONFIG_SYS_CLOCK_EXISTS)
```
to make the code more self-explanatory (`#ifdef CONFIG_SYS_CLOCK_EXISTS` is a bit cryptic).

And since the asynchronous API allows specifying timeouts, if it is to be supported even if timeouts are not available, `uart[e]_nrfx_tx` and `uart[e]_nrfx_rx_enable` should return `-ENOTSUP` (and perhaps provide some `LOG_ERR` message to make it clear what's wrong) when a timeout value other than `SYS_FOREVER_MS` is used.

nordic-krch commented:
Frankly speaking, this code looks quite nasty with dozens of ifdefs thrown here and there and no test to validate if that works.
What about trying to just wrap timer api calls with `IS_ENABLED(CONFIG_SYS_CLOCK_EXISTS)`? That should do with code looking much nicer. That will come with the slight ram cost because data related to timing will still be in the struct but it will be maintainable.

Also would be good to add test configuration in `uart_async_api` to validate it.

Can you also extract nrfx_glue changes into separate commit and PR?

anangl commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -77,6 +77,7 @@ config UART_0_NRF_HW_ASYNC
 	bool "Use hardware RX byte counting"
 	depends on UART_0_NRF_UARTE
 	depends on UART_ASYNC_API
+	depends on SYS_CLOCK_EXISTS
```
This should be done also for other instances.

anangl commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -73,8 +73,10 @@ static struct {
 	volatile size_t rx_offset;
+#ifdef CONFIG_SYS_CLOCK_EXISTS
 	int32_t rx_timeout;
 	struct k_timer rx_timeout_timer;
```
`tx_timeout_timer` should be cut out as well.

nordic-krch commented:
> As this is my first contribution to zephyr I am not sure how to properly create a test case.

You can try to extend testcase.yml in `tests/drivers/uart/uart_async_api` with the configuration where sys clock is disabled. Probably some tests will not make sense there. you can skip like
```
if (!IS_ENABLED(CONFIG_SYS_CLOCK_EXISTS`)) ztest_test_skip();
```
See https://docs.zephyrproject.org/latest/guides/test/twister.html?highlight=twister for details how to add configuration to yml file.

mspeder committed:
module: hal_nordic: fix busy_wait when system clock is disabled

When SYS_CLOCK_EXISTS is disabled, k_busy_wait is not available.
In that case this fix replaces it with nrfx_coredep_delay_us.

Signed-off-by: Matthieu Speder <mspeder@users.sourceforge.net>

mspeder committed:
drivers: serial: nrf: ignore async timeout when system clock is disabled

When SYS_CLOCK_EXISTS is disabled, k_timers functions are not available.
In that case, timeout values are simply ignored but this is the expected
behavior anyway, as mentioned in the SYS_CLOCK_EXISTS documentation.
This fix disables unused code, which would cause compilation errors
otherwise.
If SYS_CLOCK_EXISTS is enabled (which is generally the case), this fix
does not change previous behavior.

Signed-off-by: Matthieu Speder <mspeder@users.sourceforge.net>

mspeder committed:
test: serial: async uart when system clock is disabled

Validate async uart correct behavior when system clock is disabled.
In that case only tests without timeout are relevant.

Signed-off-by: Matthieu Speder <mspeder@users.sourceforge.net>

mspeder commented:
> You can try to extend testcase.yml in `tests/drivers/uart/uart_async_api` with the configuration where sys clock is disabled. Probably some tests will not make sense there. you can skip like
> 
> ```
> if (!IS_ENABLED(CONFIG_SYS_CLOCK_EXISTS`)) ztest_test_skip();
> ```
> 
> See https://docs.zephyrproject.org/latest/guides/test/twister.html?highlight=twister for details how to add configuration to yml file.

@nordic-krch I spent a few hours and multiple tentatives trying to have this work. Unfortunately it seems the test engine is not very happy when SYS_CLOCK is disabled.

```
/opt/toolchains/zephyr-sdk-0.12.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a(ztest.c.obj): in function `arch_k_cycle_get_32':
--
| /workdir/zephyr/twister-out/nrf9160dk_nrf9160/tests/drivers/uart/uart_async_api/drivers.uart.uart_async_api.noclock/../../../../../../../include/arch/arm/aarch32/misc.h:26: undefined reference to `sys_clock_cycle_get_32'
```

I did not find any test case trying to disable SYS_CLOCK anywhere else either.

So I start feeling that fixing this is beyond a reasonable scope, considering the small benefit it brings.
But before giving up on adding a test case, I'd like to have your point of view.

nordic-krch commented:
@mspeder I was affraid that test framework will relay on it and it may not be that easy to change it. Alternatively, you can try to create new test in `tests/drivers/uart`. Something like `uart_async_no_sys_clock`. Do not enable ztest there, simply enable uart async api in one instance. Add `build_only: true` property. Add readme to that test explaining that purpose is only to validate that configuration compiles and ztest does not support no sys_clock configuration.

You can also create an issue that ztest does not support no sys clock configuration.
